File: backend/app/domain/ports.py
"""依存性逆転のためのポートインターフェース（Protocolクラス）。

ドメイン層・ユースケース層はこれらのプロトコルに依存する。
アダプター層が具体的な実装を提供する。
"""

from __future__ import annotations

from typing import Protocol

from .models import GpsPoint, IntersectionResult, Route, Trip


# ---------------------------------------------------------------------------
# リポジトリポート
# ---------------------------------------------------------------------------


class TripRepository(Protocol):
    def get(self, trip_id: str) -> Trip | None: ...
    def save(self, trip: Trip) -> None: ...
    def list_all(self) -> list[Trip]: ...


class GpsRepository(Protocol):
    def get_points(self, trip_id: str) -> list[GpsPoint]: ...
    def append_points(self, trip_id: str, points: list[GpsPoint]) -> None: ...


class RouteRepository(Protocol):
    def save_route(self, trip_id: str, route: Route) -> None: ...
    def get_route(self, trip_id: str) -> Route | None: ...
    def save_intersection_results(
        self, trip_id: str, results: list[IntersectionResult],
    ) -> None: ...
    def get_intersection_results(self, trip_id: str) -> list[IntersectionResult]: ...


# ---------------------------------------------------------------------------
# ゲートウェイポート（外部サービス）
# ---------------------------------------------------------------------------


class RoutingService(Protocol):
    """ルーティングサービス"""
    async def get_bicycle_route(
        self, origin: tuple[float, float], destination: tuple[float, float],
    ) -> Route: ...

File: backend-go/internal/domain/ports.go
package domain

import "context"

// TripRepository はトリップの永続化を抽象化するインターフェース．
// インメモリ実装や将来のDB実装を差し替え可能にする．
type TripRepository interface {
	Get(tripID string) (*Trip, error)    // ID指定でトリップを取得．存在しない場合は (nil, nil)
	Save(trip *Trip) error               // トリップを保存（作成・更新兼用）
	ListAll() ([]*Trip, error)           // 全トリップを開始日時の降順で取得
}

// GpsRepository はGPSポイントの永続化を抽象化するインターフェース．
type GpsRepository interface {
	GetPoints(tripID string) ([]GpsPoint, error)              // トリップに紐づく全GPSポイントを取得
	AppendPoints(tripID string, points []GpsPoint) error      // GPSポイントをバッチ追記
}

// RouteRepository はルートと交差点判定結果の永続化を抽象化するインターフェース．
type RouteRepository interface {
	SaveRoute(tripID string, route *Route) error                            // ルートを保存（上書き）
	GetRoute(tripID string) (*Route, error)                                 // ルートを取得．未設定時は (nil, nil)
	SaveIntersectionResults(tripID string, results []*IntersectionResult) error // 交差点結果を一括保存
	GetIntersectionResults(tripID string) ([]*IntersectionResult, error)        // 交差点結果を一括取得
}

// RoutingService は外部ルーティングAPIを抽象化するインターフェース．
// 実装は OsrmGateway が担う．
type RoutingService interface {
	GetBicycleRoute(ctx context.Context, origin, destination LatLng) (*Route, error)
}

// GeocodingService はジオコーディング（地名検索）APIを抽象化するインターフェース．
// 実装は PhotonGateway が担う．
type GeocodingService interface {
	Search(ctx context.Context, query string, limit int) ([]GeocodingResult, error)
}

// RoutePlanner はルート計画ユースケースを抽象化するインターフェース．
// ハンドラー層が usecase パッケージに直接依存しないために定義する．
type RoutePlanner interface {
	Plan(ctx context.Context, tripID string, origin, destination LatLng) (*Route, error)
}

// GpsAnalyzer はGPS分析ユースケースを抽象化するインターフェース．
// GPSバッチ受信→一時停止判定→リルートの一連の処理を実行する．
type GpsAnalyzer interface {
	Execute(ctx context.Context, tripID string, points []GpsPoint) (*GpsAnalysisResult, error)
}
